crates/xtask/src/changelog.rs

```diff
@@ -0,0 +1,549 @@
+    /// acknowledge that it's already up-to-date.
+    skip_counter: u32,
+}
+
+impl Changelog {
+    fn read_file(changelog_path: &str) -> Result<Changelog> {
+        let changelog_contents = wasefire_cli_tools::fs::read(&changelog_path)?;
+
+        Self::parse(String::from_utf8(changelog_contents)?)
+    }
+
+    /// Converts raw file contents into a Changelog data structure.
+    ///
+    /// Validates requirements in docs/contributing/changelog.md
+    fn parse(contents: String) -> Result<Changelog> {
+        let mut skip_counter = 0;
+        let mut releases: Vec<Release> = Vec::new();
+
+        let mut lines = contents.lines().filter(|line| !line.is_empty()).peekable();
+
+        // First line is H1.
+        if lines.next() != Some("# Changelog") {
+            return Err(anyhow!("H1 with 'Changelog' is required"));
+        }
+
+        while let Some(mut current_line) = lines.next() {
+            let mut next_line = lines.peek();
+
+            // Parse a whole release
+            if let Some(version_string) = current_line.strip_prefix("## ") {
+                let mut release = Release::from(version_string)?;
+
+                // Parse each release type (major, minor, patch)
+                while next_line.is_some_and(|line| line.starts_with("### ")) {
+                    // Advance
+                    current_line = lines.next().unwrap();
+                    next_line = lines.peek();
+
+                    let release_type_string = current_line.trim_start_matches("### ");
+                    let release_type = ReleaseType::from_str(release_type_string, true)
+                        .map_err(|err| anyhow!("Failed to parse version_type: {err}"))?;
+
+                    let mut release_descriptions = Vec::new();
+
+                    // Parse the release descriptions
+                    // Some lines may be a new description. Some lines may be continuation of
+                    // previous descriptions.
+                    while next_line
+                        .is_some_and(|line| line.starts_with("- ") || line.starts_with("  "))
+                    {
+                        // Advance
+                        current_line = lines.next().unwrap();
+                        next_line = lines.peek();
+
+                        // New description
+                        if current_line.starts_with("- ") {
+                            release_descriptions
+                                .push(current_line.trim_start_matches("- ").to_string());
+                        } else if current_line.starts_with("  ") {
+                            // Append to previous description
+                            let mut last = release_descriptions
+                                .pop()
+                                .expect("No last description to append to.");
+                            last.push_str(format!("\n{current_line}").as_str());
+                            release_descriptions.push(last);
```

Comment:
You should be able to do this in place.

```rust
write!(release_descriptions.last_mut().context("Invalid continuation")?, "\n{current_line}");
```

Maybe not for this PR, but I think we should store logical lines and pretty-print with line folding/wrapping.

```
- This very long
  line
```

would thus become `This very long line` which would essentially be the user input to `cargo xtask changelog change`.

We should also check that descriptions don't end with a dot (can also be another PR).

Reply:
Nice, using write! now and also added validation to make sure each description does not end in a "."

Reply:
We can include whole lines and pretty print ourselves in future PR.